File: 2017-09-15/src/MyMovieCollectionSpecs.hs
import Test.Hspec
import Data.List

data Movie = Movie {
    title :: String,
    year :: Int
} deriving (Show, Eq)

matches query = isInfixOf query . title

findByTitle = filter . matches

main = hspec $ do
    let movies = [ Movie {
            title= "The Matrix",
            year= 1999
        }, Movie {
            title= "A beautiful mind",
            year= 2001
        }, Movie {
            title= "Intouchable",
            year= 2011
        }, Movie {
            title= "Forest Gump",
            year= 1994
        }]
    
    describe "My movie collection search by name" $ do
        it "should return empty when none found" $ do
            findByTitle "Interstellar" movies `shouldBe` []
        it "should return movie if title matches" $ do
            findByTitle "The Matrix" movies `shouldBe` [ Movie {
                title= "The Matrix",
                year= 1999
            }]
        it "should return all movies that matches" $ do
            findByTitle "o" movies `shouldBe` [ Movie {
                    title= "Intouchable",
                    year= 2011
                }, Movie {
                    title= "Forest Gump",
                    year= 1994
                }]

File: 2017/09/15/test/myMovieCollectionSpecs.js
"use strict";
import chai from "chai";

chai.should();

// Movie => String
let title = movie => {
    return movie.title;
};

// String => String => Bool
let isInfixOf = query => string => {
    return string.includes(query);
};

// (Movie=>Bool) => [Movie] => [Movie]
let filter = predicate => collection => {
    return collection.filter(predicate);
};

// (.)
let compose = f => g => x => f(g(x));

// String => (Movie => Bool)
let matches = query => compose(isInfixOf(query))(title);

// String => [Movie] => [Movie]
let findByTitle = compose(filter)(matches);

describe('My movie collection search by name', () => {
    let movies = [
        {
            title: "The Matrix",
            year: 1999
        },
        {
            title: "A beautiful mind",
            year: 2001
        },
        {
            title: "Intouchable",
            year: 2011
        },
        {
            title: "Forest Gump",
            year: 1994
        }
    ];
    it('should return empty when none found', () => {
        findByTitle("Interstellar")(movies).should.be.empty;
    });
    it('should return a matching movie', () => {
        findByTitle("The Matrix")(movies).should.deep.equal([{
            title: "The Matrix",
            year: 1999
        }]);
    });
    it('should return all matching movies', () => {
        findByTitle("o")(movies).should.deep.equal([{
            title: "Intouchable",
            year: 2011
        }, {
            title: "Forest Gump",
            year: 1994
        }]);
    });
});
